### PR TITLE
Destroy an ender crystal when hit by an egg.

### PR DESCRIPTION
**Pull request by bearbin**
Fixed version of #2577.

### CODE REVIEW

**Comment by SphinxC0re** [comment]
@cuberite/core-team any objections?

**Comment by tigerw** [comment]
No objections, but *how* did the branch become out of date? Nobody committed anything relevant to master.

**Comment by bearbin** [comment]
I'm not exactly sure, I think it might be that any new change to master could break the build, and you have to re-merge every time there is a new commit?

**Comment by bearbin** [comment]
Just to make sure that the PR still works and doesn't break the build.

File: src/Entities/ThrownEggEntity.cpp [hunk 1]
```diff
@@ -32,11 +32,15 @@ void cThrownEggEntity::OnHitSolidBlock(const Vector3d & a_HitPos, eBlockFace a_H
 void cThrownEggEntity::OnHitEntity(cEntity & a_EntityHit, const Vector3d & a_HitPos)
 {
 	int TotalDamage = 0;
-	// TODO: If entity is Ender Crystal, destroy it
-	
+	// If entity is an Ender Crystal, destroy it
+	if (a_EntityHit.IsEnderCrystal())
+	{
+		a_EntityHit.Destroy();
```
Comment (tigerw) [inline src/Entities/ThrownEggEntity.cpp]: Oh, perhaps a `return` here?

Comment (NiLSPACE) [inline src/Entities/ThrownEggEntity.cpp]: That means we wouldn't spawn any chickens. True, they'd explode right afterwards, but still.

Comment (tigerw) [inline src/Entities/ThrownEggEntity.cpp]: Alright.

File: src/Entities/ThrownEggEntity.cpp [hunk 2]
```diff
@@ -32,11 +32,15 @@ void cThrownEggEntity::OnHitSolidBlock(const Vector3d & a_HitPos, eBlockFace a_H
...
+	}
+
 	TrySpawnChicken(a_HitPos);
 	a_EntityHit.TakeDamage(dtRangedAttack, this, TotalDamage, 1);
```
Comment (tigerw) [inline src/Entities/ThrownEggEntity.cpp]: I wonder if this should kill the crystal anyway. It should if the crystal only has one health. Hm.

**Comment by madmaxoft** [comment]
I think the check and destruction should be inserted *after* the `TakeDamage` call. Or even better, shouldn't the `TakeDamage` call for the ender crystal destroy the crystal after any damage?

**Comment by bearbin** [comment]
It seems like the takedamage call should destroy the ender crystal anyway. Has anybody actually tested this?

If the crystal is destroyed with the old code, all that needs doing is the comment removing.

**Comment by madmaxoft** [comment]
Now waitasecond, do eggs really cause any damage in Vanilla? I thought they didn't.

**Comment by madmaxoft** [comment]
No, eggs do damage only to the enderdragon: http://minecraft.gamepedia.com/Egg#Combat

**Comment by bearbin** [comment]
OK, I will fix the PR :P

**Comment by bearbin** [comment]
The PR should be fixed now. Ender dragons and eggs are damaged when hit, and other mobs are simply knocked back.

File: src/Entities/Entity.h [hunk 3]
```diff
@@ -630,7 +630,3 @@ class cEntity
 } ;  // tolua_export
 
 typedef std::list<cEntity *> cEntityList;
-
-
-
-
```
Comment (madmaxoft) [inline src/Entities/Entity.h]: Keep the end lines

File: src/Entities/ThrownEggEntity.cpp [hunk 4]
```diff
@@ -32,11 +32,16 @@ void cThrownEggEntity::OnHitSolidBlock(const Vector3d & a_HitPos, eBlockFace a_H
 void cThrownEggEntity::OnHitEntity(cEntity & a_EntityHit, const Vector3d & a_HitPos)
 {
 	int TotalDamage = 0;
-	// TODO: If entity is Ender Crystal, destroy it
-	
+	// If entity is an Ender Dragon or Ender Crystal, it is damaged.
+	if ((a_EntityHit.IsMob() && (static_cast<cMonster &>(a_EntityHit).GetMobType() == mtEnderDragon)) ||
+		a_EntityHit.IsEnderCrystal())
```
Comment (madmaxoft) [inline src/Entities/ThrownEggEntity.cpp]: Formatting:
```cpp
if (
  (isA() && (a == b)) ||  // Is an ender dragon
  isB()
)
{
```

**Comment by bearbin** [comment]
I can't fix the issues now, I need to fix my editor for the lines at the end, but I fixed the other formatting issue.

**Comment by Seadragon91** [comment]
I would suggest deactivating the package whitespace in atom, that makes it unreadable..

**Comment by bearbin** [comment]
Whitespace is fixed now.

**Comment by SphinxC0re** [comment]
Looks good. Merge if compiles

**Pull request merged**